lib/mspec/runner/actions/constants_leak_checker.rb

```diff
@@ -0,0 +1,86 @@
+class ConstantsLeakChecker
+  attr_accessor :constants_new
+
+  def initialize
+    @constants_before = Module.constants
+  end
+
+  def check
+    @constants_after = Module.constants
+    @constants_new = @constants_after - @constants_before
+    @constants_new.empty?
+  end
+end
+
+class ConstantsLockFile
+  LOCK_FILE_NAME = '.mspec.constants.lock'
+
+  def self.exist?
+    File.exist?(LOCK_FILE_NAME)
+  end
+
+  def self.load
+    File.read(LOCK_FILE_NAME).split
+  end
+
+  def self.dump(ary)
+    File.write(LOCK_FILE_NAME, ary.map(&:to_s).uniq.sort.join("\n"))
+  end
+end
+
+class ConstantsLeakLockAction
+  def register
+    MSpec.register :start, self
+    MSpec.register :finish, self
+  end
+
+  def start
+    @checker = ConstantsLeakChecker.new
+  end
+
+  def finish
+    unless @checker.check
+      if ConstantsLockFile.exist?
+        constants_locked = ConstantsLockFile.load
+        result = constants_locked | @checker.constants_new.map(&:to_s)
+        ConstantsLockFile.dump(result)
+      else
+        ConstantsLockFile.dump(@checker.constants_new)
+      end
+    end
+  end
+end
+
+class ConstantsLeakCheckerAction
+  def register
+    MSpec.register :start, self
+    MSpec.register :finish, self
+  end
+
+  def start
+    @checker = ConstantsLeakChecker.new
+  end
+
+  def finish
```

Comment:
I was thinking to use `after` like LeakChecker so we can report which spec leaked the constant.

Then maybe seeing how similar this is to some of the LeakChecker checks, we should maybe integrate checking constants directly there.

I didn't think of extra top-level constants as "leaks" before but it makes sense to me now.

Reply:
> I was thinking to use after like LeakChecker so we can report which spec leaked the constant.

Yep. Makes sense.

> Then maybe seeing how similar this is to some of the LeakChecker checks, we should maybe integrate checking constants directly there.

Yep. Agree. So existing leaks checks will fail CI as well.

Reply:
> Yep. Agree. So existing leaks checks will fail CI as well.

Yes, that'd be good :)